### PR TITLE
New avoidance

### PR DESCRIPTION
**Pull request by gdoffe**
This PR refactors avoidance which was not working in its actual configuration due to:

- Obstacles are oriented according to sensor position. Obstacles were oriented by an hazardous supposition that the robot obstacle was oriented to rush on our robot. However it cannot be determined just with a relative distance sensor.
- Moreover as obstacle are polygons when the robot rotates on itself, the obstacle rotates too leading to crossing the avoidance path. The result is the robot oscillating in front of an obstacle.

As simulation PID parameters have been tuned for testing, also add a commit on this.

### CODE REVIEW

File: lib/obstacles/include/obstacles.h [hunk 1]
```diff
@@ -42,14 +42,15 @@ extern "C" {
...
-    uint32_t min_distance;  /**< Minimun distance from origin to create an obstacle */
-    uint32_t max_distance;  /**< Maximum distance from origin to create an obstacle */
+    uint32_t default_radius;  /**< Obstacle default radius */
+    uint32_t min_distance;      /**< Minimun distance from origin to create an obstacle */
+    uint32_t max_distance;      /**< Maximum distance from origin to create an obstacle */
 } obstacles_params_t;
 
 typedef struct {
-    pose_t points[4];       /**< Points defining the obstacle */
     double angle;           /**< Orientation of the obstacle */
```
[the remark this answers is not in the document]
Comment (gdoffe) [reply lib/obstacles/include/obstacles.h]: Nope :)

File: platforms/cortex/include/platform.h [hunk 2]
```diff
@@ -153,12 +149,13 @@
...
-#define OBSTACLE_DYN_SIZE                   400 /**< Obstacle size */
-
-#define OBSTACLE_DETECTION_MINIMUM_TRESHOLD 10  /**< Minimum obstacle detection
-                                                   treshold */
-#define OBSTACLE_DETECTION_MAXIMUM_TRESHOLD 200 /**< Maximum obstacle detection
-                                                   treshold */
+#define AVOIDANCE_RADIUS_MARGIN             0.2                     /**< Obstacle circle bounding box mark-up */
+#define LIDAR_MIN_DISTANCE                  (ROBOT_MARGIN)          /**< Minimum Lidar detection distance */
+#define LIDAR_MAX_DISTANCE                  1200                    /**< Maximum Lidar detection distance */
+#define LIDAR_MINIMUN_INTENSITY             1000                    /**< Minimum intensity required to validate a Lidar distance */
```
[the remark this answers is not in the document]
Comment (gdoffe) [reply platforms/cortex/include/platform.h:155]: It is more related to avoidance than to the robot.
Maybe we should create a new section then ?

File: lib/collisions/collisions.c [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    }
+    return res;
+}
+
+inline double collisions_compute_slope(const pose_t *a, const pose_t *b)
+{
+    return (b->y - a->y) / (b->x - a->x);
+}
+
+inline double collisions_compute_ordonate(double slope, const pose_t *b)
```
Comment (Kheprys) [inline lib/collisions/collisions.c:168]: 'ordonate' doesn't mean anything, do you mean "ordinate" ?

File: lib/collisions/include/collisions.h [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+/**
+ * @brief Compute ordonate of a line defined by two points A,B.
+ *
+ * @param[in]   a           point A
+ * @param[in]   b           point B
+ *
+ * @return                  (AB) ordonate
+ */
+double collisions_compute_ordonate(double slope, const pose_t *b);
```
Comment (Kheprys) [inline lib/collisions/include/collisions.h:149]: Same "ordonate" doesn't mean anything, do you mean "ordinate" ?

**Review by Kheprys — COMMENTED** [review]
Interresting changes. I've made a quick review, i'll need more time to review it in a more detailed way.

**Comment by gdoffe** [comment]
> **collisions module:** since it's a module, its public functions should have a module prefix.
> 
> **collisions/avoidance:** I don't like using the `pose_t` type everywhere.
> This type is for a robot position (ie X, Y and angle). In these modules, it's only 2D points, so `vector_t` should be more appropriate, but the name is far from perfect.
> 
> Some commits still need to be squashed, like the "[Fix](https://github.com/cogip/mcu-firmware/pull/39/commits/cd89d84279876fc0a1aa69284d8f20a45718e090)" commit, but I would also squash " [use circle](b3bd757da194a74c6df6e7d5ebca58815601cee6)" and "[add multi-forms obstacles](https://github.com/cogip/mcu-firmware/pull/39/commits/5b34835e331529cc009393fcd557e08430ca15da)" commits, since the second mostly modifies the former one, and makes review difficult.

done

**Review by ecourtois — CHANGES_REQUESTED** [review]
In [commit](https://github.com/cogip/mcu-firmware/pull/39/commits/da6739c83aa9e9705f33592de8fd4a7525f4796a) message:
- '**OBSTACLE_RECTANGLE**' type is based on new 'rectangle_t' type.
- not signed

In [commit](https://github.com/cogip/mcu-firmware/pull/39/commits/7c4c70237569a77432340812a635e78d17a9e1cc) message:
- "It also changes return types to better **suit** needs."
- Remove conflict comments

In [commit](https://github.com/cogip/mcu-firmware/pull/39/commits/fcf9b28cdf219585277bdd74c6b5153a9a5c992a) message:
- "Avoidance should not care about the **obstace** type."
- Remove conflict comments

**Review by ecourtois — APPROVED** [review]
LGTM!!!

**Pull request merged**